grid/tree-view-docs (#24535)
highcharts/highcharts: docs/grid/rows/tree-view.md

---

> sidebar_label: "Tree view"

Reviewer: ```suggestion
```

Not needed when same as first #

tags: ["grid-pro"]
---

# Tree view

Tree view is a Highcharts Grid Pro feature that turns flat row data into an

> expandable hierarchy inside the Grid. Use it when your data represents

Reviewer: ```suggestion
expandable hierarchy inside the grid. Use it when your data represents
```
Never capital G in grid when using definite noun phrase

parent-child relationships such as organization structures, folders, product
categories, account trees, or multi-level budgets.

> TreeView is configured in `data.treeView` and works with the local/client-side

Reviewer: ```suggestion
Tree view is configured using `data.treeView` and works with the local
```

data provider. Grid builds a tree index from the source table, then projects
the queried row set into visible parent and child rows before pagination is
applied.

> ## Requirements

Reviewer: ```suggestion
## Minimum requirements
```

- Tree view is available in Highcharts Grid Pro.

> - Set `data.idColumn` to a column containing stable string or number row IDs.

Reviewer: Is this also needed when `input.type: 'path'` ?

Author: At this stage, yes

- Tree view works with local data (`data.columns`,
  `data.dataTable`, or connector-backed data).

```js
Grid.grid('container', {
    data: {
        idColumn: 'id',
        treeView: {
            treeColumn: 'name'
        }
    }

> });

Reviewer: For clarity I think this snippet should show a full, bare minimum required config, also containing data.columns.path. treeView shouldonly have strictly needed props, and use defaults instead.

Then we expand in the samples below for more "advanced" cases.

```

## Input models

> TreeView supports two input models:

Reviewer: ```suggestion
Tree view supports two input models:
```

            expandedRowIds: 'all'
        }
    },
    header: ['name', 'budget']
});
```

`parentIdColumn` defaults to `parentId`.

### `path` input

Use `path` input when hierarchy is stored as full path strings or when you want
Grid to infer missing intermediate ancestors automatically.

```js
Grid.grid('container', {
    data: {
        columns: {
            id: [1, 2, 3, 4, 5],
            path: [
                'Root',
                'Root/Sales',
                'Root/Marketing',
                'Root/Sales/EMEA',
                'Root/Sales/APAC'
            ],
            budget: [1000, 600, 400, 350, 250]
        },
        idColumn: 'id',
        treeView: {
            input: {
                type: 'path',
                pathColumn: 'path',
                separator: '/',
                showFullPath: true
            }
        }
    }
});
```

`pathColumn` defaults to `path`, `separator` defaults to `'/'`, and
`showFullPath` defaults to `false`.

### Generated ancestors for path input

With `path` input, intermediate nodes do not need to exist as source rows. If
your dataset contains `Root/Sales/EMEA` but does not contain separate source

> rows for `Root` or `Root/Sales`, TreeView can generate those ancestors

Reviewer: ```suggestion
rows for `Root` or `Root/Sales`, Tree view can generate those ancestors
```

automatically so the visible hierarchy stays complete.

Generated rows are structural rows. Their path value is available, but other
columns are rendered as `null` unless you provide explicit source rows for
those ancestors.

## Tree column

The expand/collapse UI is rendered in the tree column.

- Use `data.treeView.treeColumn` to choose which column receives indentation
  and toggle buttons.
- If `treeColumn` is omitted, Grid uses the first rendered column.
- For path input, `showFullPath: false` renders only the current segment when
  the path column is also the tree column. Set it to `true` to render the full
  path string.

```js
treeView: {
    treeColumn: 'name',
    input: {
        type: 'path',
        showFullPath: false
    }
}
```

## Initial expansion state

Use `expandedRowIds` to seed which branches are expanded when TreeView is

> initialized or when TreeView configuration changes.

Reviewer: ```suggestion
Use `expandedRowIds` to seed which branches are expanded when Tree view is
initialized or when Tree view configuration changes.
```

Whe should perhaps also mention that `expandedRowIds` uses column defined in `idColumn`?

```js
treeView: {
    expandedRowIds: [1, 2, 7] // or: 'all'
}
```

`expandedRowIds: 'all'` expands every row that currently has children. Leaf
rows are ignored automatically.

## Sticky parents

Set `stickyParents: true` to keep the current ancestor context visible while
scrolling long trees.

```js
treeView: {
    stickyParents: true
}
```

Sticky parents are enabled by default. This pairs especially well with deep
hierarchies and [row virtualization](https://www.highcharts.com/docs/grid/rows/virtualization).

## Runtime API

> The TreeView controller is available on `grid.treeView`.

Reviewer: ```suggestion
The Tree view controller is available on `grid.treeView`.
```

```js
await grid.treeView?.expandAll();
await grid.treeView?.collapseAll();
await grid.treeView?.toggleRow(2);
```

Use this API for toolbar actions, external buttons, or synchronized UI outside

> the Grid.

Reviewer: ```suggestion
the grid.
```

## Events

> TreeView exposes row toggle lifecycle events at the root `events` option:

Reviewer: ```suggestion
Tree view exposes row toggle lifecycle events at the root `events` option:
```

- `beforeTreeRowToggle`
- `afterTreeRowToggle`

`beforeTreeRowToggle` can cancel the interaction by calling
`event.preventDefault()`.

```js
Grid.grid('container', {
    events: {
        beforeTreeRowToggle: function (event) {
            if (event.rowId === 1) {
                event.preventDefault();
            }
        },
        afterTreeRowToggle: function (event) {
            console.log(event.rowId, event.expanded);
        }
    }
});
```

Both events include:

- `rowId`
- `expanded`
- `originalEvent` when the toggle came from mouse or keyboard interaction

## Sorting, filtering, and pagination

> TreeView is applied after sorting and filtering, and before pagination. In

Reviewer: ```suggestion
Tree view is applied after sorting and filtering, and before pagination. In
```

practice this means:

- sorting changes the visible sibling order inside the tree
- filtering can keep matching descendants and reintroduce required ancestors so
  the visible result still forms a valid hierarchy
- pagination counts projected visible tree rows, not only raw source rows

> This behavior makes TreeView work naturally with Grid querying without breaking

Reviewer: ```suggestion
This behavior makes Tree view work naturally with Grid querying without breaking
```

parent-child context.

## Custom separators

For path data that is not slash-separated, pass a string, `RegExp`, or callback
in `data.treeView.input.separator`.

```js
treeView: {
    input: {
        type: 'path',
        separator: /[A-Z]+(?![a-z])|[A-Z][a-z]*/
    }
}
```

> This is useful for camel-cased hierarchy keys such as

Reviewer: ```suggestion
This is useful for e.g. PascalCase hierarchy keys such as
```
